challenges/backend/src/main/java/com/headstorm/interview/backend/DataManager.java

```diff
@@ -0,0 +1,72 @@
+package com.headstorm.interview.backend;
+
+import java.util.Arrays;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+public class DataManager
+{
+	private static final ReentrantReadWriteLock LOCK = new ReentrantReadWriteLock();
```

Comment:
I'm impressed that you thought ahead to lock your mutable data structures for access by multiple threads of execution. Unfortunately, this style of programming is very defensive and in larger systems will force you to anticipate every possible failure point and synchronize.  Assuming you are not allowed to use Locks or Synchronized access, can you elaborate on different pattern you would use for managing mutable data structures for concurrent computation that would also allow you to use multiple threads of execution?

*This is a potentially difficult question. Since we're less interested in your knowledge of a large number of various concurrent programming paradigms and more interested in your ability to apply and describe new patterns to existing problems, the clue is 'Actor Pattern'.  Exited to read your response!

Reply:
If you wanted to use an Actor Pattern as opposed to lock-based concurrency, the REST webserver end would be able to remain largely the same. As opposed to locking and modifying the state directly, the web calls would instead send messages to an Actor responsible for managing the numerical data. Since this REST api has been designed and implemented in Java, let's use Akka as our Actor library. We can use TELL messages for the PUT and PATCH webserver actions in order to update the numerical data the API returns, and we can use ASK messages for the GET webserver action to return data after any preceding TELL messages have been processed. This is where the power of the REST webserver library we're using will be helpful, as we can actually instruct the webserver to complete the web call based on the CompletableFuture obtained when using Akka's ASK message type.